File: 112. Path Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        if not root:
            return False
        self.res = False
        self.DFS(root, sum)
        return self.res
    
    def DFS(self, root, sum):
        if root == None:
            return
        if root.left == None and root.right == None:
            if sum == root.val:
                self.res = True
        self.DFS(root.left, sum - root.val)
        self.DFS(root.right, sum - root.val)

File: 112. Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if (root == NULL) return false;
        bool res = false;
        DFS(root, sum, res);
        return res;
    }
    
    void DFS(TreeNode* root, int sum, bool &res)
    {
        if (root == NULL)
            return;
        if (root->left == NULL && root->right == NULL)
        {
            if (sum == root->val)
                res = true;
        }
        DFS(root->left, sum - root->val, res);
        DFS(root->right, sum - root->val, res);
    }
};
